espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/espresso/EspressoW3CActionAdapter.kt

```diff
@@ -0,0 +1,158 @@
+package io.appium.espressoserver.lib.helpers.w3c.adapter.espresso
+
+import android.content.Context
+import android.graphics.Point
+import android.util.DisplayMetrics
+import android.view.View
+
+import java.util.Collections
+
+import androidx.test.espresso.UiController
+import androidx.test.espresso.action.GeneralLocation
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
+import io.appium.espressoserver.lib.helpers.AndroidLogger
+import io.appium.espressoserver.lib.helpers.Logger
+import io.appium.espressoserver.lib.helpers.w3c.adapter.BaseW3CActionAdapter
+import io.appium.espressoserver.lib.helpers.w3c.dispatcher.W3CKeyEvent
+import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType
+import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState
+import io.appium.espressoserver.lib.model.Element
+
+import android.view.KeyEvent.ACTION_DOWN
+import android.view.KeyEvent.ACTION_UP
+import android.view.MotionEvent.ACTION_MOVE
+import android.view.MotionEvent.ACTION_POINTER_DOWN
+import android.view.MotionEvent.ACTION_POINTER_UP
+import androidx.test.core.app.ApplicationProvider.getApplicationContext
+import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.Helpers.isTouch
+import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.Helpers.toCoordinates
+
+class EspressoW3CActionAdapter(private val uiController: UiController) : BaseW3CActionAdapter() {
+
+    private val androidKeyEvent: AndroidKeyEvent
+    private val multiTouchState = MultiTouchState()
+    private val displayMetrics = getApplicationContext<Context>().getResources().getDisplayMetrics()
+
+    override val viewportHeight: Long
+        get() = displayMetrics.heightPixels.toLong()
+
+    override val viewportWidth: Long
+        get() = displayMetrics.widthPixels.toLong()
+
+    override val logger: Logger
+        get() = AndroidLogger.logger
+
+    init {
+        this.androidKeyEvent = AndroidKeyEvent(uiController)
+    }
+
+    @Throws(AppiumException::class)
+    override fun keyDown(keyEvent: W3CKeyEvent) {
+        androidKeyEvent.keyDown(keyEvent)
+    }
+
+    @Throws(AppiumException::class)
+    override fun keyUp(keyEvent: W3CKeyEvent) {
+        androidKeyEvent.keyUp(keyEvent)
+    }
+
+    @Throws(AppiumException::class)
+    override fun pointerDown(button: Int?, sourceId: String?, pointerType: PointerType?,
+                    x: Float?, y: Float?, depressedButtons: Set<Int>,
+                    globalKeyInputState: KeyInputState?) {
+        this.logger.info(String.format("Running pointer down at coordinates: %s %s %s", x, y, pointerType))
```

Comment:
I would rather use template string everywhere

Reply:
Agreed. Pushed the fix. I really wish the translator did that for us.